### PR TITLE
Add support for location lineage post processing

### PR DESCRIPTION
**Pull request by lincmba**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Resolves [link to issue]

**Engineer Checklist**

- [x] I have written **Unit tests** for any new feature(s) and edge cases for
      bug fixes
- [ ] I have added documentation for any new feature(s) and configuration
      option(s) on the `README.md`
- [x] I have run `mvn spotless:check` to check my code follows the project's
      style guide
- [x] I have run `mvn clean test jacoco:report` to confirm the coverage report
      was generated at `plugins/target/site/jacoco/index.html`
- [x] I ran `mvn clean package` right before creating this pull request.

### CODE REVIEW

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java [hunk 1]
```diff
@@ -247,9 +249,41 @@ public String postProcess(RequestDetailsReader request, HttpResponse response)
             resultContent = this.fhirR4JsonParser.encodeResourceToString(practitionerDetailsBundle);
         }
 
+        String resourceType = request.getResourceName();
+
+        if (Constants.SyncStrategy.LOCATION.equals(resourceType)) {
+            String method = request.getRequestType().name();
+            if ("POST".equals(method) || "PUT".equals(method)) {
```
Comment (ndegwamartin) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java:256]: I think we should merge these two conditions into one:

```suggestion
        if (Constants.SyncStrategy.LOCATION.equals(resourceType) && ("POST".equals(request.getRequestType().name()) || "PUT".equals(request.getRequestType().name()))) { 
```

File: plugins/src/test/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecisionTest.java [hunk 2]
```diff
@@ -997,6 +1000,76 @@ public void testPostProcessWithListModeHeaderSearchByTagPaginateEntriesBundle()
...
+            locationHelperMock
+                    .when(() -> LocationHelper.updateLocationLineage(any(), any()))
+                    .thenReturn(location);
+
+            testInstance.postProcess(requestDetailsSpy, fhirResponseMock);
+            locationHelperMock.verify(
+                    () -> LocationHelper.updateLocationLineage(eq(iGenericClient), eq("123")));
+
+            Assert.assertNotNull(location);
+            Assert.assertEquals("Location/123", location.getId());
```
Comment (ndegwamartin) [inline plugins/src/test/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecisionTest.java:1044]: These asserts are redundant for the unit tests given the values are being set by Line 1022 and 1023. We should remove them.

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java [hunk 3]
```diff
@@ -247,9 +249,41 @@ public String postProcess(RequestDetailsReader request, HttpResponse response)
...
+        String locationId;
+        if (StringUtils.isNotBlank(resultContent)) {
+            IBaseResource parsedResource = this.fhirR4JsonParser.parseResource(resultContent);
+            if (parsedResource instanceof Location) {
+                return ((Location) parsedResource).getIdElement().getIdPart();
+            }
+        }
+
+        String[] pathParts = requestPath.split("/");
+        locationId = pathParts[pathParts.length - 1];
```
[the remark this answers is not in the document]
Comment (lincmba) [reply plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java:278]: They're not for differentiating the request method. They are there incase a request is made but the resource updated/created is not returned in the json. Done away with one condition though

**Review by ndegwamartin — APPROVED** [review]
